core/src/main/scalanative/sttp/client3/AbstractCurlBackend.scala

```diff
@@ -57,45 +55,73 @@ abstract class AbstractCurlBackend[F[_]](monad: MonadError[F], verbose: Boolean)
       }
 
       val spaces = responseSpace
-      curl.option(WriteFunction, AbstractCurlBackend.wdFunc)
-      curl.option(WriteData, spaces.bodyResp)
-      curl.option(HeaderData, spaces.headersResp)
-      curl.option(Url, request.uri.toString)
-      curl.option(TimeoutMs, request.options.readTimeout.toMillis)
-      setMethod(curl, request.method)
-
-      setRequestBody(curl, request.body)
-
-      responseMonad.flatMap(lift(curl.perform)) { _ =>
-        curl.info(ResponseCode, spaces.httpCode)
-        val responseBody = fromCString((!spaces.bodyResp)._1)
-        val responseHeaders_ = parseHeaders(fromCString((!spaces.headersResp)._1))
-        val httpCode = StatusCode((!spaces.httpCode).toInt)
-
-        if (headers.ptr != null) headers.ptr.free()
-        multiPartHeaders.foreach(_.ptr.free())
-        free((!spaces.bodyResp)._1)
-        free((!spaces.headersResp)._1)
-        free(spaces.bodyResp.asInstanceOf[Ptr[CSignedChar]])
-        free(spaces.headersResp.asInstanceOf[Ptr[CSignedChar]])
-        free(spaces.httpCode.asInstanceOf[Ptr[CSignedChar]])
-        curl.cleanup()
-
-        val statusText = responseHeaders_.head.name.split(" ").last
-        val responseHeaders = responseHeaders_.tail
-        val responseMetadata = ResponseMetadata(httpCode, statusText, responseHeaders)
-
-        val body: F[T] = bodyFromResponseAs(request.response, responseMetadata, Left(responseBody))
-        responseMonad.map(body) { b =>
-          Response[T](
-            body = b,
-            code = httpCode,
-            statusText = statusText,
-            headers = responseHeaders,
-            history = Nil,
-            request = request.onlyMetadata
-          )
-        }
+      BinaryFileHelper.getFilePath(request.response) match {
+        case Some(file) =>
+          val outputPath = file.toPath.toString
+          val outputFilePtr: Ptr[FILE] = fopen(toCString(outputPath), toCString("wb"))
+          curl.option(WriteData, outputFilePtr)
+          curl.option(Url, request.uri.toString)
+          setMethod(curl, request.method)
+          setRequestBody(curl, request.body)
+          responseMonad.flatMap(lift(curl.perform)) { _ =>
+            curl.info(ResponseCode, spaces.httpCode)
+            val httpCode = StatusCode((!spaces.httpCode).toInt)
+            if (headers.ptr != null) headers.ptr.free()
+            multiPartHeaders.foreach(_.ptr.free())
+            free(spaces.httpCode.asInstanceOf[Ptr[CSignedChar]])
+            fclose(outputFilePtr)
+            curl.cleanup()
+            val responseMetadata = ResponseMetadata(httpCode, "", List.empty)
```

Comment:
can't we get the response headers here?

Reply:
nope as they were written with `curl.option(HeaderData, spaces.headersResp)` which we cannot set with current setting

Reply:
at least with reasonable effort it may be possible somehow with `HEADERFUNCTION` and `HEADERDATA`  but it will require creation of another `wdFunc`-like function and I am not sure if it will work